extension/src/time_series/pipeline/fill_to.rs

```diff
@@ -0,0 +1,248 @@
+                let right_wt = 1. - (rhs.ts - target_ts) as f64 / interval;
+                TSPoint{ts: target_ts, val: lhs.val * left_wt + rhs.val * right_wt}
+            },
+            FillToMethod::Nearest =>
+                if rhs.ts - target_ts >= target_ts - lhs.ts {
+                    TSPoint{ts: target_ts, val: lhs.val}
+                } else {
+                    TSPoint{ts: target_ts, val: rhs.val}
+                },
+        }
+    }
+}
+
+// TODO is (immutable, parallel_safe) correct?
+#[pg_extern(
+    immutable,
+    parallel_safe,
+    name="fill_to",
+    schema="toolkit_experimental"
+)]
+pub fn fillto_pipeline_element<'e> (
+    fill_method: String,
+    interval: Interval,
+) -> toolkit_experimental::UnstableTimevectorPipeline<'e> {
+    unsafe {
+        let interval = interval as *const pg_sys::Interval;
+        if (*interval).day > 0 || (*interval).month > 0 {
+            panic!("downsample intervals are currently restricted to stable units (hours or smaller)");
+        }
+        let interval = (*interval).time;
+
+        let fill_method = match fill_method.to_lowercase().as_str() {
+            "locf" => FillToMethod::LOCF,
+            "interpolate" => FillToMethod::Interpolate,
+            "linear" => FillToMethod::Interpolate,
+            "nearest" => FillToMethod::Nearest,
+            _ => panic!("Invalid fill method")
+        };
+
+        Element::FillTo {
+            interval,
+            fill_method
+        }.flatten()
+    }
+}
+
+pub fn fill_to<'s>(
+    series: &toolkit_experimental::Timevector<'s>,
+    element: &toolkit_experimental::Element
+) -> toolkit_experimental::Timevector<'s> {
+    let (interval, method) = match element {
+        Element::FillTo{interval, fill_method} => (*interval, fill_method),
+        _ => unreachable!()
+    };
+
+    match series.series {
+        SeriesType::ExplicitSeries{..} => panic!("Timeseries must be sorted prior to passing to fill_to"),
+        SeriesType::NormalSeries{step_interval, ..} => if step_interval <= interval { return series.clone(); },
+        _ => ()
+    }
+
+    let mut result: Option<Vec<TSPoint>> = None;
+    let mut it = series.iter().peekable();
+    let mut current = it.next();
+
+    while current.is_some() {
+        let lhs = current.unwrap();
+        let rhs = it.peek();
+
+        if rhs.is_some() && rhs.unwrap().ts - lhs.ts > interval {
+            let rhs = rhs.unwrap();
+            if result.is_none() {
+                result = Some(series.iter().collect());
+            }
+            let result = result.as_mut().unwrap();
+            let mut target = lhs.ts + interval;
+            while target < rhs.ts {
+                result.push(method.fill_point(&lhs, &rhs, target));
+                target += interval;
+            }
+        }
+
+        current = it.next();
+    }
+
+    if result.is_some() {
+        let mut result = result.unwrap();
+        result.sort_unstable_by_key(|p| p.ts);
```

Comment:
1. do we need this? is this guaranteed to be sorted by the fact that we have sorted input?
2. Unless we deduplicate, I think we should use a stable sort, that we if there's meaning to the current order we'll preserve it
```suggestion
        result.sort_by_key(|p| p.ts);
```

Reply:
> is this guaranteed to be sorted...
Oh, I see: we push the new elements onto the end of the vector

Reply:
fixed